graphql/src/Mu/GraphQL/Query/Parse.hs

```diff
@@ -184,192 +193,245 @@ instance ParseArgs p '[] where
 instance (KnownName aname, ParseArg p a, ParseArgs p as, FindDefaultArgValue aanns)
          => ParseArgs p ('ArgSingle ('Just aname) aanns a ': as) where
   parseArgs vmap args
-    = case find ((== nameVal (Proxy @aname)) . T.unpack . GQL.unName . GQL._aName) args of
+    = let aname = T.pack $ nameVal (Proxy @aname)
+      in case find ((== nameVal (Proxy @aname)) . T.unpack . GQL.unName . GQL._aName) args of
         Just (GQL.Argument _ x)
-          -> (:*) <$> (ArgumentValue <$> parseArg' vmap x)
+          -> (:*) <$> (ArgumentValue <$> parseArg' vmap aname x)
                   <*> parseArgs vmap args
-        Nothing -> case findDefaultArgValue (Proxy @aanns) of
-                     Just x  -> (:*) <$> (ArgumentValue <$> parseArg' vmap (constToValue x))
-                                     <*> parseArgs vmap args
-                     Nothing -> empty
+        Nothing
+          -> do x <- findDefaultArgValue (Proxy @aanns) aname
+                (:*) <$> (ArgumentValue <$> parseArg' vmap aname (constToValue x))
+                     <*> parseArgs vmap args
 
 class FindDefaultArgValue (vs :: [Type]) where
-  findDefaultArgValue :: Alternative f
+  findDefaultArgValue :: MonadError T.Text f
                       => Proxy vs
+                      -> T.Text
                       -> f GQL.ValueConst
 instance FindDefaultArgValue '[] where
-  findDefaultArgValue _ = empty
+  findDefaultArgValue _ aname
+    = throwError $ "argument '" <> aname <> "' was not given a value, and has no default one"
 instance {-# OVERLAPPABLE #-} FindDefaultArgValue xs
          => FindDefaultArgValue (x ': xs) where
   findDefaultArgValue _ = findDefaultArgValue (Proxy @xs)
 instance {-# OVERLAPS #-} ReflectValueConst v
          => FindDefaultArgValue (DefaultValue v ': xs) where
-  findDefaultArgValue _ = pure $ reflectValueConst (Proxy @v)
+  findDefaultArgValue _ _ = pure $ reflectValueConst (Proxy @v)
 
-parseArg' :: (ParseArg p a, Alternative f)
+parseArg' :: (ParseArg p a, MonadError T.Text f)
```

Comment:
So basically when you want to fail with a message, you `throwError` and rely on `MonadError` instead of Alternative|Applicative? Seems useful! 🤔

Reply:
Yep, that's also why having `Alternative f` is useful, you just replace it by `MonadError e` and the compiler tells you where to change things.

One problem with this approach is that you can only throw *one* error for the entire parsing, whereas in some cases you may want to give as many errors as possible.